File: scripts/process-gharok-sprites.mjs
import sharp from 'sharp';
import { copyFileSync, mkdirSync, existsSync, statSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');
const SRC_DIR = join(ROOT, 'assets', 'gharok');
const RAW_DIR = join(SRC_DIR, '_raw');
const SIZE = 512;

async function processOne(name) {
  const src = join(SRC_DIR, `${name}.png`);
  const raw = join(RAW_DIR, `${name}.png`);
  mkdirSync(RAW_DIR, { recursive: true });
  if (!existsSync(raw)) copyFileSync(src, raw);

  const { data, info } = await sharp(raw).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
  const { width: w, height: h } = info;
  const px = data;

  const sample = (x, y) => {
    const i = (y * w + x) * 4;
    return [px[i], px[i + 1], px[i + 2]];
  };
  const samples = [sample(2, 2), sample(w - 3, 2), sample(2, h - 3), sample(w - 3, h - 3),
    sample(10, 10), sample(w - 11, 10)];
  const br = Math.round(samples.reduce((a, s) => a + s[0], 0) / samples.length);
  const bg = Math.round(samples.reduce((a, s) => a + s[1], 0) / samples.length);
  const bb = Math.round(samples.reduce((a, s) => a + s[2], 0) / samples.length);

  const out = Buffer.alloc(w * h * 4);
  let minX = w, minY = h, maxX = 0, maxY = 0;
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const i = (y * w + x) * 4;
      const r = px[i], g = px[i + 1], b = px[i + 2];
      const dr = r - br, dg = g - bg, db = b - bb;
      const dist = Math.sqrt(dr * dr + dg * dg + db * db);
      const mx = Math.max(r, g, b), mn = Math.min(r, g, b);
      const isLightNeutral = mx >= 230 && (mx - mn) <= 18;
      const isSoftBg = mx >= 210 && (mx - mn) <= 12 && dist < 40;
      let a;
      if (dist < 22 || isLightNeutral || isSoftBg) a = 0;
      else if (dist < 45) a = Math.round(255 * (dist - 22) / 23);
      else a = 255;
      out[i] = r; out[i + 1] = g; out[i + 2] = b; out[i + 3] = a;
      if (a > 16) {
        if (x < minX) minX = x;
        if (y < minY) minY = y;
        if (x > maxX) maxX = x;
        if (y > maxY) maxY = y;
      }
    }
  }
  if (maxX < minX) throw new Error('empty after key: ' + name);

  const pad = 8;
  minX = Math.max(0, minX - pad);
  minY = Math.max(0, minY - pad);
  maxX = Math.min(w - 1, maxX + pad);
  maxY = Math.min(h - 1, maxY + pad);
  const cw = maxX - minX + 1, ch = maxY - minY + 1;
  const cropped = Buffer.alloc(cw * ch * 4);
  for (let y = 0; y < ch; y++) {
    const srcOff = ((minY + y) * w + minX) * 4;
    out.copy(cropped, y * cw * 4, srcOff, srcOff + cw * 4);
  }

  const scale = Math.min(SIZE / cw, SIZE / ch);
  const nw = Math.max(1, Math.round(cw * scale));
  const nh = Math.max(1, Math.round(ch * scale));
  const left = Math.floor((SIZE - nw) / 2);
  const top = SIZE - nh - 4;

  const resized = await sharp(cropped, { raw: { width: cw, height: ch, channels: 4 } })
    .resize(nw, nh, { kernel: 'lanczos3' })
    .png()
    .toBuffer();

  await sharp({
    create: { width: SIZE, height: SIZE, channels: 4, background: { r: 0, g: 0, b: 0, alpha: 0 } },
  })
    .composite([{ input: resized, left, top }])
    .png({ compressionLevel: 9 })
    .toFile(join(SRC_DIR, `${name}.png`));

  const meta = await sharp(join(SRC_DIR, `${name}.png`)).metadata();
  console.log(`${name}: ${meta.width}x${meta.height} hasAlpha=${meta.hasAlpha} bytes=${statSync(join(SRC_DIR, `${name}.png`)).size}`);
}

mkdirSync(join(ROOT, 'www', 'assets', 'gharok'), { recursive: true });
for (const name of ['idle', 'windup', 'walk']) {
  await processOne(name);
  copyFileSync(join(SRC_DIR, `${name}.png`), join(ROOT, 'www', 'assets', 'gharok', `${name}.png`));
}
console.log('synced to www/assets/gharok');

File: scripts/process-gharok-sprites.py
"""Remove light backgrounds, trim, and pack Gharok sprites for the game."""
from PIL import Image
import os
import shutil

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SRC_DIR = os.path.join(ROOT, 'assets', 'gharok')
# Keep originals from generation under _raw if not already
RAW_DIR = os.path.join(SRC_DIR, '_raw')


def process(src, dst, size=512):
    im = Image.open(src).convert('RGBA')
    px = im.load()
    w, h = im.size
    samples = [px[2, 2], px[w - 3, 2], px[2, h - 3], px[w - 3, h - 3],
               px[10, 10], px[w - 11, 10]]
    br = sum(s[0] for s in samples) // len(samples)
    bg = sum(s[1] for s in samples) // len(samples)
    bb = sum(s[2] for s in samples) // len(samples)
    out = Image.new('RGBA', (w, h), (0, 0, 0, 0))
    opx = out.load()
    for y in range(h):
        for x in range(w):
            r, g, b, a = px[x, y]
            dr, dg, db = abs(r - br), abs(g - bg), abs(b - bb)
            dist = (dr * dr + dg * dg + db * db) ** 0.5
            mx = max(r, g, b)
            mn = min(r, g, b)
            is_light_neutral = mx >= 230 and (mx - mn) <= 18
            is_soft_bg = mx >= 210 and (mx - mn) <= 12 and dist < 40
            if dist < 22 or is_light_neutral or is_soft_bg:
                opx[x, y] = (0, 0, 0, 0)
            elif dist < 45:
                alpha = int(255 * (dist - 22) / 23)
                opx[x, y] = (r, g, b, alpha)
            else:
                opx[x, y] = (r, g, b, 255)
    bbox = out.getbbox()
    if not bbox:
        raise SystemExit('empty after key: ' + src)
    pad = 8
    l, t, r, b = bbox
    l = max(0, l - pad)
    t = max(0, t - pad)
    r = min(w, r + pad)
    b = min(h, b + pad)
    cropped = out.crop((l, t, r, b))
    cw, ch = cropped.size
    scale = min(size / cw, size / ch)
    nw, nh = max(1, int(cw * scale)), max(1, int(ch * scale))
    resized = cropped.resize((nw, nh), Image.Resampling.LANCZOS)
    canvas = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    # Bias feet toward bottom of frame for consistent ground anchor
    canvas.paste(resized, ((size - nw) // 2, size - nh - 4), resized)
    canvas.save(dst, 'PNG', optimize=True)
    c = canvas.getpixel((0, 0))
    mid = canvas.getpixel((size // 2, int(size * 0.55)))
    print(f'{os.path.basename(dst)}: {canvas.size} corner={c} midA={mid[3]} '
          f'bytes={os.path.getsize(dst)}')


def main():
    os.makedirs(RAW_DIR, exist_ok=True)
    for name in ['idle', 'windup', 'walk']:
        src = os.path.join(SRC_DIR, f'{name}.png')
        raw = os.path.join(RAW_DIR, f'{name}.png')
        if not os.path.exists(raw):
            shutil.copy2(src, raw)
        process(raw, src, 512)
    # sync into www for Capacitor
    www = os.path.join(ROOT, 'www', 'assets', 'gharok')
    os.makedirs(www, exist_ok=True)
    for name in ['idle', 'windup', 'walk']:
        shutil.copy2(os.path.join(SRC_DIR, f'{name}.png'),
                     os.path.join(www, f'{name}.png'))
    print('synced to www/assets/gharok')


if __name__ == '__main__':
    main()
